### ISSUE TITLE
Add Maven artifacts to central mvnrepository

### ISSUE THREAD

**Issue by tal-franji** [tensorflow/tensorflow#6926]
For Java/Scala users, specifically - those using the Spark/Hadoop environment - it would help to have two artifacts:
example/feature compiled protos

TFRecordFileFormat for Hadoop/Spark io

**Comment by asimshankar** [comment]
We will be looking into those, particularly for the [Java API](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java) and all that is needed for it.

However, might be a while. 

FYI @jhseu

**Comment by tal-franji** [comment]
Thanks for the uodate.
I'm willing to contributes some time for this, if this may help
--franji

**Comment by dazraf** [comment]
Don't forget that you can deploy and reference native binaries. As an example see [Netty's native capability](http://netty.io/wiki/forked-tomcat-native.html).

**Comment by asimshankar** [comment]
Starting with release 1.1, this will be done. I'll close this issue out when I update the documentation (Java README). But in the mean time, note that you can use the release candidate version by adding:

```xml
<dependency>
  <groupId>org.tensorflow</groupId>
  <artifactId>tensorflow</artifactId>
  <version>1.1.0-rc0</version>
</dependency>
```

to your project's `pom.xml`.

**Comment by kyu-sz** [comment]
@asimshankar Is the JNI library GPU-enabled or CPU-only?

**Comment by asimshankar** [comment]
The JNI library contained in the Maven package is CPU only.

However, the way the code is setup is that it looks for the `tensorflow_jni` library in the system path  (see [`NativeLibrary.java`](https://github.com/tensorflow/tensorflow/blob/7303b63/tensorflow/java/src/main/java/org/tensorflow/NativeLibrary.java#L48)). Thus, to enable GPU support, download the JNI archive mentioned in the README and make that available to the JVM (e.g., via `-Djava.library.path`)

**Comment by astrung** [comment]
can i use this maven artifact on windows. Thus, native dynamic library ( libtensorflow.so )can not be used on windows, so how can i build it without native library ?

**Comment by asimshankar** [comment]
Yes, you can use it on Windows. There was a slight issue with RC0 on Windows, but you can use the following:

```xml
 <dependency>
   <groupId>org.tensorflow</groupId>
   <artifactId>tensorflow</artifactId>
   <version>1.1.0-rc0-windows-fix</version>
 </dependency>
 ```

**Comment by dazraf** [comment]
Hi,

Like the netty native library, the convention with Maven is to use the `classifier` tag to specify OS specific distros. 
See [here](http://netty.io/wiki/forked-tomcat-native.html#wiki-h2-3) for an example.

You can even use the classifier to distinguish between pure-java implementations and those that come packaged with GPU specific binaries.

I may have missed something, but it seems like we will need to install the GPU native libraries manually. 

If so, can be please avoid this!

As developers, we should be able to integrate TensorFlow with a single `<dependency>` declaration, such that any platform specificity is fully resolved. e.g. 

```xml
<dependency>
  <groupId>org.tensorflow</groupId>
  <artifactId>tensorflow</artifactId>
  <version>1.1.0</version>
  <classifier>windows-gpu</classifier>
</dependency>
```

1. It's really easy to integrate TensorFlow
2. All build aspects, including native libraries are fully prepackaged, expanded to filesystem, and bound, accordingly at runtime.
3. The entire build process on a 'clean' machine (i.e. no TensorFlow libraries installed on path) is easily possible - which makes it easy to integrate with CI servers.

This is all standard stuff for Maven and all Java devs would expect it!

kind regards
Fuzz

**Comment by asimshankar** [comment]
@dazraf : A single Maven dependency is all that is needed (irrespective of OS).  The same:

```xml
<dependency>
   <groupId>org.tensorflow</groupId>
   <artifactId>tensorflow</artifactId>
   <version>1.1.0-rc0-windows-fix</version>
</dependency>
```

Work for Linux and OS X and when the 1.1 release is finalized, using `<version>1.1.0</version>` will work for all three OSes as well. So I'm not quite sure I follow your concern.

That said, there are some caveats:

- No, no GPU support included in the pom yet. But this is something we'll look into for the future. Till then, the workaround would be to download the GPU JNI library and make that available in the system path.
- All the platforms are packaged together, without the use of classifiers. This makes the download a little larger. I was looking into classifiers but came across this: http://maven.40175.n5.nabble.com/How-to-deploy-with-classifier-tp5523009p5523417.html Admittedly. I'm not a Maven expert so wasn't sure how to interpret that.

**Comment by dazraf** [comment]
My concerns are exactly your caveats :-)
* The classifiers minimise the size of the distro
* They keep everything in the maven "world" - native libraries and JVM classes. Which means that builds can be fully controlled through maven (without manual processes for installation of libraries).

Am I allowed to submit a PR for this? I can help fix this.

**Comment by asimshankar** [comment]
@dazraf : You're welcome to send in a PR, though I'd like to make sure I understand the proposed flow. (FYI: The PR would operate somewhere under [`tensorflow/java/maven`](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java/maven)).

To help me understand the proposed flow, consider the following scenario: User X authors a library libX that depends on TensorFlow. X does not know what platforms libX will be used on, since that choice would be made by applications that depend on libX. In this case, what happens? Does User X have to define platform-specific classifiers for libX and force libX clients to select one? Does this keep propagating? Or are end applications required to select the TensorFlow classifier even though they have no direct dependency on TensorFlow (perhaps they have to add the `os-maven-plugin` for OS but still need to explicitly choose between CPU and GPU) - which may be fine, but is more than a 4 line addition to their dependencies? (I'm guessing this is why the thread I linked to above discourages use of classifiers for different "flavors", but perhaps I'm misinterpreting)

Note that even as-is, everything is in the maven "world" - builds can be fully controlled through maven with the GPU caveat. It seems to me that including GPU support in the `.jar` for `libtensorflow_jni` is somewhat orthogonal to the use of classifiers?

I would be tempted to keep the dependency as simple as possible (e.g., "add these 4 lines, that's it"), even if advanced users (for whom the size of the distro is a real concern) have to do something extra.

Looking at the [netty link you pointed to](http://netty.io/wiki/forked-tomcat-native.html#artifacts), this seems similar to the approach used by netty where they provide the `netty-tcnative-boringssl-static` artifact for the common simple use case and suggest that more nuanced users use a different artifact/classifier. Is your proposal intended to address these "advanced" users without affecting the simpler 4-line-addition-to-pom.xml flow?

**Comment by dazraf** [comment]
@asimshankar : I see your point. So adapting what I wrote:

1. as a libX developer, I want to bind to a single dependency that allows me to write my logic. 
2. as a user of libX, in another java project, I would like to reference libX and also add a dependency to leverage my GPU. 

Do these make sense?
The model you proposed works perfectly for `1.`. A single reference `org.tensorflow:tensorflow:<version>` to bind to is nice and easy.

For `2.`, I'm looking for a maven packaging of the GPU specific libraries - something like `org.tensorflow:tensorflow-windows-gpu:<version>` such that at runtime, `tensorflow` discovers `tensorflow-windows-gpu`, unpacks it, and binds to its native lib.

This is how the `netty-native` libraries work - netty looks for a well-known class on the classpath, which is provided by the `-native-` library. Netty invokes this class, which in turn unpacks the binary (it not already present) in a well known location etc ..

So in summary, the proposition is that Java developers of `2.` would prefer the installation of the GPU library to be controllable by maven.

Is that a clearer narrative?

**Comment by asimshankar** [comment]
Clearer, but pardon me, I'm not there just yet :)

Ignoring GPU for a moment, the way the packaging currently works is that [`NativeLibrary.java`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/NativeLibrary.java) finds the appropriate native lib from the classpath, unpacks it and loads it.

In principle, the GPU-enabled JNI libraries could also be packaged the same way with a change to `NativeLibrary.java` so that it checks for GPU support and looks for the corresponding native library. Is that what you had in mind?

And then presumably, the top level application can choose to use GPU or not by adding a dependency on the `libtensorflow_jni-gpu` artifact or something?

**Comment by EronWright** [comment]
@asimshankar thanks for figuring out how to embed the native lib, it really saves a lot of hassle.

**Comment by EronWright** [comment]
One little nuance to be aware of going forward.   In hosted environments like Apache Flink where each 'job' runs in a separate classloader, numerous instances of the JNI library might simultaneously become loaded.  The fact that the extracted resource has a generated temp name makes this possible; otherwise the JVM would complain "Library xyz already loaded in another classloader".  I'm hopeful that side-by-side loading won't have a negative effect on TF.
